src/shared/messaging/test/port-rpc-test.js

```diff
@@ -175,23 +175,158 @@ describe('PortRPC', () => {
+    sender.connect(port1);
+
+    // Invoke "close" manually. In a real app it will be invoked when the
+    // window is unloaded and/or the PortRPC is destroyed.
+    sender.call('close');
+    sender.call('close');
+
+    await waitForMessageDelivery();
+
+    assert.calledOnce(closeHandler);
+  });
+
+  /** Transfer a MessagePort to another frame and return the transferred port. */
+  async function transferPort(port, targetWindow) {
+    const transferredPort = new Promise(resolve => {
+      targetWindow.addEventListener('message', e => {
+        if (e.ports[0]) {
+          resolve(e.ports[0]);
+        }
+      });
+    });
+    targetWindow.postMessage({}, '*', [port]);
+    return transferredPort;
+  }
+
+  describe('Safari <= 15 workaround', () => {
+    const safariUserAgent =
+      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.2 Safari/605.1.15';
+
+    let removeWorkaround;
+    let childFrame;
+
+    beforeEach(() => {
+      childFrame = document.createElement('iframe');
+      document.body.append(childFrame);
+
+      removeWorkaround = installPortCloseWorkaroundForSafari(safariUserAgent);
+    });
+
+    afterEach(() => {
+      removeWorkaround();
+      childFrame.remove();
+    });
+
+    it('transfers port to parent frame and sends "close" event from there when window is unloaded', async () => {
+      const { port1, port2 } = new MessageChannel();
+      const transferredPort = await transferPort(
+        port1,
+        childFrame.contentWindow
+      );
+
+      const sender = new PortRPC({
+        userAgent: safariUserAgent,
+        currentWindow: childFrame.contentWindow,
+      });
+      const receiver = new PortRPC();
+      const closeHandler = sinon.stub();
+
+      receiver.on('close', closeHandler);
+      receiver.connect(port2);
+      sender.connect(transferredPort);
       await waitForMessageDelivery();
-      assert.notCalled(testMethod);
+
+      closeHandler.resetHistory();
+
+      // Emulate the Safari bug by disabling `postMessage` on the sending port
+      // in the original frame. When the port is transferred to the "parent"
+      // frame, it will reconstituted as a new MessagePort instance.
+      transferredPort.postMessage = () => {};
+
+      // Unload the child frame. The "unload" handler will transfer the port to
+      // the parent frame and the "close" event will be sent from there.
+      childFrame.remove();
+
+      await waitForMessageDelivery();
+
+      assert.called(closeHandler);
+    });
+  });
+
+  [
+    // Chrome 100
+    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4867.0 Safari/537.36',
+
+    // Firefox 96
+    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:96.0) Gecko/20100101 Firefox/96.0',
+  ].forEach(userAgent => {
+    it('does not use workaround in unaffected browsers', () => {
+      sinon.stub(window, 'addEventListener');
```

Comment:
Shouldn't a sinon `sandbox` would be better here? So that `sandbox.restore` is called on `afterEach`?

Reply:
I agree, except that the `sinon` object is itself a sandbox in modern Sinon versions so these mocks can be reset with `sinon.restore()` in a suite-level `afterEach`.